src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,393 @@
+    bool split, double &log_ratio_prior_prob, double &log_ratio_likelihoods) {
+  /* In the function, we treat the two clusterings as the "divided" one, where
+   * first_random_idx and second_random_idx are in different clusters, and as
+   * the "united" one, where the two datapoints are in the same cluster.
+   * Only in the end we control, via the variable "split", which configuration
+   * is the original one and which one is the proposal, in order to compute
+   * correctly the returned values.
+   */
+  std::vector<std::shared_ptr<AbstractHierarchy>> divided_clust_unique_values =
+      {unique_values[0]->clone(), unique_values[0]->clone()};
+  std::shared_ptr<AbstractHierarchy> united_clust_unique_values(
+      unique_values[0]->clone());
+  Eigen::VectorXd pred_lpdf_divided_clust = Eigen::VectorXd::Zero(2);
+  double pred_lpdf_united_clust = 0;
+
+  /* We treat log_ratio_prior_prob as in the split case, then we change its
+   * sign at the end if we are in the merge case.
+   */
+  log_ratio_prior_prob = mixing->get_mass_new_cluster(
+      allocations.size(), true, true, unique_values.size());
+  log_ratio_likelihoods = 0;
+
+  std::vector<unsigned int> random_idxs = {first_random_idx,
+                                           second_random_idx};
+
+  for (unsigned int clust_idx = 0; clust_idx <= 1; ++clust_idx) {
+    pred_lpdf_divided_clust(clust_idx) +=
+        divided_clust_unique_values[clust_idx]->prior_pred_lpdf(
+            data.row(random_idxs[clust_idx]));
+    divided_clust_unique_values[clust_idx]->add_datum(
+        random_idxs[clust_idx], data.row(random_idxs[clust_idx]),
+        update_hierarchy_params());
+
+    /* FIXME@Mario: conditional_pred_lpdf should execute prior_pred_lpdf when
+     * card == 0
+     */
+    if (united_clust_unique_values->get_card() == 0) {
+      pred_lpdf_united_clust += united_clust_unique_values->prior_pred_lpdf(
+          data.row(random_idxs[clust_idx]));
+    } else {
+      pred_lpdf_united_clust +=
+          united_clust_unique_values->conditional_pred_lpdf(
+              data.row(random_idxs[clust_idx]));
+      log_ratio_prior_prob -= mixing->get_mass_existing_cluster(
+          allocations.size(), true, true, united_clust_unique_values);
+    }
+    united_clust_unique_values->add_datum(random_idxs[clust_idx],
+                                          data.row(random_idxs[clust_idx]),
+                                          update_hierarchy_params());
+
+    std::set<int> set_to_cycle;
+    if (split) {
+      set_to_cycle = restricted_gs_unique_values[clust_idx]->get_data_idx();
+      set_to_cycle.erase(random_idxs[clust_idx]);
+    } else {
+      set_to_cycle =
+          unique_values[allocations[random_idxs[clust_idx]]]->get_data_idx();
+    }
+    set_to_cycle.erase(random_idxs[clust_idx]);
+    auto it = set_to_cycle.cbegin();
+    const auto end = set_to_cycle.cend();
+
+    for (; it != end; ++it) {
```

Comment:
If you don't need the two iterators outside of the for loop, please declare them in the `()`